File: hackerrank/AngryProfessor.hs
-- https://www.hackerrank.com/challenges/angry-professor/problem

import           Text.Printf

data Answer = YES | NO deriving (Show)

data Case = Case { totalStudents         :: Int
                 , cancellationThreshold :: Int
                 , arrivals              :: [Int] }

toPairs :: [a] -> [(a, a)]
toPairs []                  = []
toPairs [_]                 = error "list size is odd, should be even" -- Break the program on erroneous input
toPairs (first:second:tail) = (first, second) : toPairs tail

toCase :: (String, String) -> Case
toCase (firstLine, secondLine) = do
  Case { totalStudents = students
       , cancellationThreshold = threshold
       , arrivals = map read $ words secondLine :: [Int] }
  where firstList = map read $ words firstLine :: [Int]
        students = head firstList
        threshold = firstList !! 1

validate :: Case -> Case
validate (Case students threshold arrivals)
  | students == length arrivals = Case students threshold arrivals
  | otherwise = error errorMessage -- Break the program on erroneous input
  where errorMessage = printf "Total students expected to be %s, but is %s" (show students) (show $ length arrivals)

countEarly :: [Int] -> Int
countEarly arrivals = length $ filter (<= 0) arrivals

isClassCancelled :: Case -> Answer
isClassCancelled (Case _ cancellationThreshold arrivals)
  | countEarly arrivals >= cancellationThreshold = NO
  | otherwise = YES

solve :: Case -> Answer
solve = isClassCancelled . validate

main =
  interact $                    -- Reads an input, prints an output
  unlines .                     -- Join elements with \n. Ex: unlines ["YES", "NO", "YES"] -> "YES\nNO\nYES\n"
  map (show . solve . toCase) . -- For each element do: 1) toCase, 2) solve, 3) show
  toPairs .                     -- Each test case has two lines, put tme in a pair: (line 1, line 2)
  tail .                        -- Drop the first element (we dont need it)
  lines                         -- Read lines from stdin

File: hackerrank/AngryProfessor.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// https://www.hackerrank.com/challenges/angry-professor/problem
public class AngryProfessor {

    private static final String SPACE = " ";

    enum Answer {
        YES, NO
    }

    private class Case {

        private final int totalStudents;
        private final int cancellationThreshold;
        private final int[] arrivals;

        public Case(int totalStudents, int cancellationThreshold, int[] arrivals) {
            this.totalStudents = totalStudents;
            this.cancellationThreshold = cancellationThreshold;
            this.arrivals = arrivals;
        }

        private void validate() {
            if (totalStudents != arrivals.length) {
                String errorMessage = String.format("Total students expected to be %s, but is %s",
                    totalStudents, arrivals.length);
                throw new AssertionError(errorMessage); // Break the program on erroneous input
            }
        }

        private int countEarly() {
            int studentsOnTime = 0;
            for (int arrival : arrivals) {
                studentsOnTime += (arrival <= 0) ? 1 : 0;
            }
            return studentsOnTime;
        }

        private Answer isClassCancelled() {
            int studentsOnTime = countEarly();
            return (studentsOnTime >= cancellationThreshold) ? Answer.NO : Answer.YES;
        }

        public Answer solve() {
            validate();
            return isClassCancelled();
        }
    }

    private Case readCase(BufferedReader reader) throws IOException {
        String[] split = reader.readLine().split(SPACE);
        int totalStudents = Integer.parseInt(split[0]);
        int cancellationThreshold = Integer.parseInt(split[1]);
        String[] arrivalsStr = reader.readLine().split(SPACE);
        int[] arrivals = Arrays.stream(arrivalsStr).mapToInt(Integer::parseInt).toArray();
        return new Case(totalStudents, cancellationThreshold, arrivals);
    }

    private List<Answer> solve() throws IOException {
        List<Answer> answers = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            int totalCases = Integer.parseInt(reader.readLine());
            for (int i = 0; i < totalCases; i++) {
                Case aCase = readCase(reader);
                answers.add(aCase.solve());
            }
        }
        return answers;
    }

    public static void main(String[] args) throws IOException {
        AngryProfessor angryProfessor = new AngryProfessor();
        List<Answer> answers = angryProfessor.solve();
        for (Answer answer : answers) {
            System.out.println(answer);
        }
    }
}
